src/main/scala/midas/widgets/LoadMem.scala

```diff
@@ -38,64 +107,45 @@ class LoadMemWidget(hKey: Field[NastiParameters], maxBurst: Int = 8)(implicit p:
 
   val wAddrH = genWOReg(Wire(UInt(max(0,  p(hKey).addrBits - 32).W)), "W_ADDRESS_H")
   val wAddrL = genWOReg(Wire(UInt(min(32, p(hKey).addrBits     ).W)), "W_ADDRESS_L")
-  val wLen = genWORegInit(Wire(UInt((p(hKey).addrBits - log2Ceil(hWidth/8) + 1).W)), "W_LENGTH", 0.U)
+  val wLen = Wire(Decoupled(UInt(p(hKey).addrBits.W)))
   // When set, instructs the unit to write 0s to the complete address space
   // Cleared when completed
-  val zeroOutDram = genAndAttachReg(Wire(Bool()), "ZERO_OUT_DRAM", Some(false.B))
-
-  val wBeatsRemaining = RegInit(0.U(log2Ceil(maxBurst+1).W))
-  val nextBurstLen = Mux(wLen > maxBurst.U, maxBurst.U, wLen)
-  val wAddr = Cat(wAddrH, wAddrL)
-
-  when (zeroOutDram && wLen === 0.U) {
-    when (wBeatsRemaining === 0.U) {
-      // Commence initialization by faking a really large write
-      wLen := 1.U << (p(hKey).addrBits - log2Ceil(hWidth/8))
-      wAddrH := 0.U
-      wAddrL := 0.U
-    }.elsewhen(io.toSlaveMem.w.fire && io.toSlaveMem.w.bits.last) {
-      // We've written the last beat; clear the zeroOutDram bit to indicate doneness
-      zeroOutDram := false.B
-    }
-  }
-
-  io.toSlaveMem.aw.bits := NastiWriteAddressChannel(
-      id = 0.U,
-      addr = wAddr,
-      len = nextBurstLen - 1.U,
-      size = size)(p.alterPartial({ case NastiKey => p(hKey) }))
+  val zeroOutDram = Wire(Decoupled(Bool()))
 
-  io.toSlaveMem.aw.valid := wLen > 0.U && wBeatsRemaining === 0.U
+  attachDecoupledSink(wLen, "W_LENGTH")
+  attachDecoupledSink(zeroOutDram, "ZERO_OUT_DRAM")
 
-  val nextWAddr = wAddr + (nextBurstLen << (log2Ceil(hWidth/8).U))
-
-  when (io.toSlaveMem.aw.fire) {
-    wLen := wLen - nextBurstLen
-    wBeatsRemaining := nextBurstLen
-    if (p(hKey).addrBits > p(CtrlNastiKey).dataBits) {
-      wAddrH := nextWAddr(p(hKey).addrBits - 1, 32)
-      wAddrL := nextWAddr(31, 0)
-    } else {
-      wAddrL := nextWAddr(p(hKey).addrBits - 1, 0)
-    }
-  }
+  val hAlterP = p.alterPartial({ case NastiKey => p(hKey) })
+  val wAddrQ = Module(new Queue(new LoadMemWriteRequest()(hAlterP), 2))
+  wAddrQ.io.enq.valid := wLen.valid
+  wAddrQ.io.enq.bits.zero := false.B
+  wAddrQ.io.enq.bits.addr := Cat(wAddrH, wAddrL)
+  wAddrQ.io.enq.bits.len  := wLen.bits
+  wLen.ready := wAddrQ.io.enq.ready
 
   val wDataQ = Module(new MultiWidthFifo(cWidth, hWidth, maxBurst))
   attachDecoupledSink(wDataQ.io.in, "W_DATA")
 
-  io.toSlaveMem.w.bits := NastiWriteDataChannel(
-    data = Mux(zeroOutDram, 0.U, wDataQ.io.out.bits),
-    last = wBeatsRemaining === 1.U
-  )(p alterPartial ({ case NastiKey => p(hKey) }))
-
-  when (io.toSlaveMem.w.fire) {
-    wBeatsRemaining := wBeatsRemaining - 1.U
-  }
-  io.toSlaveMem.w.valid := (zeroOutDram || wDataQ.io.out.valid) && wBeatsRemaining =/= 0.U
-  wDataQ.io.out.ready := !zeroOutDram && io.toSlaveMem.w.ready && wBeatsRemaining =/= 0.U
-
-  // TODO: Handle write responses better?
-  io.toSlaveMem.b.ready := Bool(true)
+  val extMem = p(ExtMem).get
```

Comment:
This is probably going to break some of the midas examples that may not set this field. I think the safer thing to do is to still just zero the entire host address space.